1789 Implement recommendations from the GradebookNG accessibility audit

payten opened the pull request:
Delivers #1789.

Hi @steveswinsburg, sorry, kind of a large PR.

The big changes in here are custom implementations of the `DataTable` and `HeaderToolbar` as `GbDataTable` and `GbHeaderToolbar` respectively.  I didn't really want to subclass `DataTable`, but had trouble applying the generics required for the `newCellItem` override!  Any tips?  I've also customised the HTML for DataTable, adding the row role in there.. but that could be moved to an override of `newRowItem` if the generics can be fudged!

One trick I had to pull was passing the assignments and categories to the `GbHeaderToolbar`.  At the moment these are passed to the `GbDataTable` and stored there.  I wonder if these could become final attributes on the `GradebookPage` and made accessible via a public getter?  I use these in GbHeaderToolbar to build the categories row.

Cheers,
Payten

payten commented on gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/component/GbDataTable.java:
```diff
@@ -0,0 +1,49 @@
+
+	@Override
+	protected Item<IColumn<T, S>> newCellItem(final String id, final int index, final IModel<IColumn<T, S>> model) {
+		return new Item(id, index, model) {
```
So this is the method it'd be nice to override inline as per `new DataTable() {@Override protected Item<IColumn> newCellItem(......` if only I could figure out the generics :confounded:

steveswinsburg replied:
If we could do this would this mean we don't need the GbDataTable?

payten commented on gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/component/GbDataTable.java:
```diff
@@ -0,0 +1,49 @@
+public class GbDataTable<T, S> extends DataTable<T, S> {
+
+	IModel<Map<String, Object>> gradebookData;
+
```
Ah.. this is the `gradebookData` I pass in.  Maybe could be accessible from `((GradebookPage) getPage()).getAssignments()` and ((GradebookPage) getPage()).getCategoryDefinitions()?

steveswinsburg replied:
I think this would be ok. So long as the assignments and categorydefinitions were wrapped in methods and weren't stored in the GradebookPage itself (as variables), as then that page would get massif.

payten replied:
Hm.. so if assignments and categorydefinitions are not stored on the GradebookPage, would we still be able to reuse the assignments already returned from the service?  Or would the getter simply hit the service again?

steveswinsburg replied:
It would have to hit the service. this is only called once on construction though right? Does the GradebookPage even need the assignments/category definitions if all of the work is occuring in here? Could they just be passed in?

payten replied:
Yep, I'm happy to leave them as passed in via the model as there'll be no extra service calls.  I was expecting we'd get rid of `GbDataTable` and move `newCellItem` override line on in GradebookPage.java like  `new DataTable {@Override... newCellItem...}`.  The only thing holding me back from doing that was the generics and compilation errors when I tried to get it working.

payten commented on gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/component/GbHeadersToolbar.java:
```diff
@@ -0,0 +1,115 @@
+	public void onInitialize() {
+		super.onInitialize();
+
+		final Map<String, Object> gradebookData = ((GbDataTable)getTable()).getGradebookData();
```
And here is where I use the `gradebookData` as the toolbar knows its DataTable I can grab them.  This component also knows its page... so maybe?

steveswinsburg replied:
Right so this could just get the data from the GradebookPage directly ey

payten commented on gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/pages/GradebookPage.java:
```diff
@@ -388,7 +389,15 @@ public String getCssClass() {
+					currentColumnIndex = currentColumnIndex + category.getAssignmentList().size();
+					cols.add(currentColumnIndex, column);
+					currentColumnIndex = currentColumnIndex + 1;
+				} else {
```
Oo.. I replaced that really ugly IColumn sort and inline comparator with something a little simpler.. we can use the assignment counts to add the category average column at the right point in the columns array.  Hurray!

payten commented on gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/pages/GradebookPage.java:
```diff
@@ -399,65 +408,15 @@ public String getCssClass() {
-				}
-			});
-		}
-
```
Laters ugly comparator.

steveswinsburg replied:
:walking: :door:

payten replied:
Oops.  You'll thank me.. one day.

steveswinsburg replied:
That was the ugly comparator leaving with he door hitting it on the way out.
Github needs a _like_ button.

payten commented on gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/pages/GradebookPage.java:
```diff
@@ -399,65 +408,15 @@ public String getCssClass() {
 		// TODO make this AjaxFallbackDefaultDataTable
-		final DataTable table = new DataTable("table", cols, studentGradeMatrix, 100);
+		final GbDataTable table = new GbDataTable("table", cols, studentGradeMatrix, 100, Model.ofMap(modelData));
 		table.addBottomToolbar(new NavigationToolbar(table) {
```
Hm.. grossly passing in extra data to the custom DataTable.. feeling more like this should be available from the GradebookPage...

steveswinsburg replied:
yeah I think if this was a getter on the page this would be ok.

payten commented on gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/panels/AddOrEditGradeItemPanel.html:
```diff
@@ -5,9 +5,6 @@
-		
-			<h2 wicket:id="heading">Add/Edit Title Here</h2>
-
             <div wicket:id="addGradeFeedback"></div>
```
All headers have been removed from the panels and are set on the window itself.  Thankfully all of these panels are only ever used in the context of a window.

payten commented on gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/panels/GradebookSpreadsheetFixedTables.html:
```diff
@@ -1,28 +0,0 @@
-<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
```
Now that categories are rendered by the `GbToolbarHeader` I can remove both `GradebookSpreadsheetFixedTables.java` and `GradebookSpreadsheetFixedTables.html`... they were kinda lame anyway.

steveswinsburg replied:
Oh man I modified that yesterday! :stuck_out_tongue_closed_eyes:

steveswinsburg commented:
OK, big one. Let me have a look and I'll get back to you. I'll merge in the others after a once over. This one is in conflict btw :(

steveswinsburg commented on gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/pages/GradebookPage.java:
```diff
@@ -390,7 +393,15 @@ public String getCssClass() {
-				cols.add(column);
+				if (settings.isCategoriesEnabled()) {
+					// insert category column after assignments in that category
+					currentColumnIndex = currentColumnIndex + category.getAssignmentList().size();
```
Does this work ok for a category with no assignments? NPE? or is getAssignmentList() always set?

payten replied:
Surprisingly yes! Because the assignments are already sorted by their category, all the uncategorised columns are at the end of the array.  This only iterates over the real categories and pokes the category average column into the right spot of the already sorted cols array.

steveswinsburg replied:
OK nice one.

payten commented:
Okidokes! My brain figured out the generics stuff finally.. so I've dropped `GbDataTable` and now only override the methods we need to inline.  I've also updated `GbHeaderToolbar` to take the model data it requires to build the extra category header row.  I think that should solve the issues raised in the conversation above.  But let me know there's anything else I can do to improve it..  Thanks!
